src/tools/bpf_link_helper/bpf-link-helper.c

```diff
@@ -0,0 +1,282 @@
+
+  /* started nested attribute for XDP */
+  nla = (struct nlattr *)(((char *)&req) + NLMSG_ALIGN(req.nh.nlmsg_len));
+  nla->nla_type = NLA_F_NESTED | IFLA_XDP;
+  nla->nla_len = NLA_HDRLEN;
+
+  /* add XDP fd */
+  nla_xdp = (struct nlattr *)((char *)nla + nla->nla_len);
+  nla_xdp->nla_type = IFLA_XDP_FD;
+  nla_xdp->nla_len = NLA_HDRLEN + sizeof(int);
+  memcpy((char *)nla_xdp + NLA_HDRLEN, &prog_fd, sizeof(prog_fd));
+  nla->nla_len += nla_xdp->nla_len;
+
+  flags = XDP_FLAGS_SKB_MODE;
+  nla_xdp = (struct nlattr *)((char *)nla + nla->nla_len);
+  nla_xdp->nla_type = IFLA_XDP_FLAGS;
+  nla_xdp->nla_len = NLA_HDRLEN + sizeof(flags);
+  memcpy((char *)nla_xdp + NLA_HDRLEN, &flags, sizeof(flags));
+  nla->nla_len += nla_xdp->nla_len;
+
+  req.nh.nlmsg_len += NLA_ALIGN(nla->nla_len);
+
+  if( send(sock, &req, req.nh.nlmsg_len, 0) < 0 ) {
+    ret = -1;
+    goto cleanup;
+  }
+
+  ret = netlink_recv(sock, nl_pid, seq);
+
+cleanup:
+  close(sock);
+  return ret;
+}
+
+static int
+__bpf_prog_get_next_id(__u32 start_id, __u32 *next_id)
+{
+  union bpf_attr attr;
+  int err;
+
+  memset(&attr, 0, sizeof(attr));
+  attr.start_id = start_id;
+
+  err = sys_bpf(BPF_PROG_GET_NEXT_ID, &attr, sizeof(attr));
+  if( !err )
+    *next_id = attr.next_id;
+
+  return err;
+}
+
+static int
+__bpf_prog_get_fd_by_id(__u32 id)
+{
+  union bpf_attr attr;
+
+  memset(&attr, 0, sizeof(attr));
+  attr.prog_id = id;
+
+  return sys_bpf(BPF_PROG_GET_FD_BY_ID, &attr, sizeof(attr));
+}
+
+static int
+__bpf_obj_get_info_by_fd(int bpf_fd, void *info, size_t info_len)
+{
+  union bpf_attr attr;
+
+  memset(&attr, 0, sizeof(attr));
+  attr.info.bpf_fd = bpf_fd;
+  attr.info.info_len = (__u32)info_len;
+  attr.info.info = (__u64)info;
+
+  return sys_bpf(BPF_OBJ_GET_INFO_BY_FD, &attr, sizeof(attr));
+}
+
+int main(int argc, char *argv[])
+{
+  __u32 id = 0;
+  char *ifname;
+  unsigned int ifindex;
+  char *prog_name;
+  struct stat stat;
+
+  /* Redirect output to kernel buffer */
+  if( fstat(STDOUT_FILENO, &stat) != 0 ) {
+    int fd = open("/dev/kmsg", O_WRONLY);
+    dup2(fd, STDOUT_FILENO);
+    dup2(fd, STDERR_FILENO);
+    close(fd);
+  }
+
+  if( argc != 3 && argc != 2 ) {
+    fprintf(stderr, "Usage: %s ifname [prog_name]\n", argv[0]);
+    return -1;
+  }
+
+  ifname = argv[1];
+  prog_name = argc == 3 ? argv[2] : NULL;
+
+  ifindex = if_nametoindex(ifname);
+
+  if( prog_name == NULL ) {
+    int rc;
+
+    printf("Unlinking from %s\n", ifname);
+    rc = xdp_link_netlink(-1, ifindex);
+    if( rc != 0 )
+      fprintf(stderr, "Failed to unlink\n");
+    return rc;
+  }
+
+  while( true ) {
+    int fd;
+    struct bpf_prog_info info = {};
+    int err = __bpf_prog_get_next_id(id, &id);
+
+    if (err) {
+      if (errno == ENOENT) {
+        break;
+      }
+      fprintf(stderr, "can't get next program: %s%s\n", strerror(errno),
+              errno == EINVAL ? " -- kernel too old?" : "");
+      return -1;
+    }
+
+    fd = __bpf_prog_get_fd_by_id(id);
```

Comment:
Shall we close the file descriptor later in the function, e.g. if the program names don't match and we carry on to the next one? This is mainly for correctness unless you think the application's lifespan is too short to worry about such things.

Reply:
Done